sorc/ncep_post.fd/CALCHIPSI.f

```diff
@@ -0,0 +1,154 @@
+!> @file
+!> @brief Subroutine that computes the velocity potential and
+!> streamfunction from isobaric winds.
+!>
+!><pre>
+!> This routine is based on the CFS genpsiandchi program that
+!> computes velocity potential and streamfunction from the
+!> isobaric wind components. The program was authored and provided
+!> by Saha and H. Chuang.
+!> Given the U-V wind components at P-points, this routine
+!> collects the winds in the full IM,JM,LSM domain,
+!> transforms them back to spectrum space and computes divergence,
+!> vorticity, streamfunction and potential. The routine returns: 
+!> 	  PSI: the streamfunction in global domain
+!> 	  CHI: the velocity potential in global domain
+!></pre>
+!>   
+!> @param[in] UISO real U-wind component (m/s) at all P-points.
+!> @param[in] VISO real V-wind component (m/s) at all P-points.
+!> @param[out] CHI real velocity potential (m^2/s) in full grid domain at all P-points.
+!> @param[out] PSI real streamfunction (m^2/s) in full grid domain at all P-points
+!>
+!> ### Program history log:
+!> Date       | Programmer   | Comments
+!> -----------|--------------|---------
+!> 2024-07-17 | Karina Asmar | Initial   
+!> 2024-07-25 | Jesse Meng   | Fixes for mpi processes 
+!>
+!> @author Karina Asmar EMC/VPPPG @date 2024-07-17
+!-----------------------------------------------------------------------
+!> @brief Subroutine that computes velocity potential and streamfunction
+!> from isobaric winds.  
+!>   
+!> @param[in] UISO real U-wind component (m/s) at all P-points.
+!> @param[in] VISO real V-wind component (m/s) at all P-points.
+!> @param[out] CHI real velocity potential (m^2/s) in full grid domain at P-points.
+!> @param[out] PSI real streamfunction (m^2/s) in full grid domain at P-points
+!-----------------------------------------------------------------------
+      SUBROUTINE CALCHIPSI(UISO,VISO,CHI,PSI)
+!
+!     INCLUDE ETA GRID DIMENSIONS.  SET/DERIVE OTHER PARAMETERS.
+!     
+      use gridspec_mod, only: IDRT
+      use ctlblk_mod, only: ISTA, IEND, JSTA, JEND, IM, JM, LSM, ME, SPVAL, MPI_COMM_COMP,&
+                            num_procs, icnt, idsp, isxa, iexa, jsxa, jexa
+!
+!- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+      implicit none
+
+      include 'mpif.h'
+!
+!     DECLARE VARIABLES.
+!     
+      integer :: JCAP, I, J, L, IERR
+      REAL, dimension(ISTA:IEND,JSTA:JEND,LSM), intent(in) :: UISO, VISO
+
+      real, dimension(IM,JM,LSM) :: IN_UWIND, IN_VWIND, OUT_UWIND, OUT_VWIND, DIV, ZO, CHI_OUT, PSI_OUT, CHI, PSI
```

Comment:
@KarinaAsmar-NOAA @JesseMeng-NOAA You might consider using allocable variables and reducing the number of variables in the full domain to to conserve memory.

Reply:
@WenMeng-NOAA I'll try with allocatable variables and let you know.